test/Transactions/Orleans.Transactions.Tests/Runners/TransactionRecoveryTestsRunner.cs

```diff
@@ -0,0 +1,147 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Microsoft.Extensions.Logging;
+using Orleans.Runtime;
+using Orleans.TestingHost;
+using Xunit;
+using Xunit.Abstractions;
+using Microsoft.Extensions.DependencyInjection;
+using Orleans.Providers;
+using Orleans.TestingHost.Utils;
+using Orleans.Hosting;
+
+namespace Orleans.Transactions.Tests
+{
+    public class TransactionRecoveryTestsRunner : TransactionTestRunnerBase
+    {
+        private readonly Random seed;
+        private readonly TestCluster testCluster;
+        private readonly ILogger logger;
+        private static TimeSpan RecoveryTimeout = TimeSpan.FromSeconds(60);
+        public TransactionRecoveryTestsRunner(TestCluster testCluster, ITestOutputHelper output)
+            : base(testCluster.GrainFactory, output)
+        {
+            this.testCluster = testCluster;
+            this.logger = this.testCluster.ServiceProvider.GetService<ILogger<TransactionRecoveryTestsRunner>>();
+            this.seed = new Random();
+        }
+
+        public class SiloBuilderConfigurator : ISiloBuilderConfigurator
+        {
+            public void Configure(ISiloHostBuilder hostBuilder)
+            {
+                hostBuilder.ConfigureServices(services => services.AddTransientNamedService<IControllable, TMGrainLocator>(typeof(TMGrainLocator).Name));
+            }
+        }
+
+        public virtual async Task TransactionWillRecoverAfterRandomSiloFailure(string transactionTestGrainClassName, bool killSiloWhichRunsTm)
+        {
+            const int grainCount = 100;
+            const int opsPerGrain = 2;
+            var txGrains = Enumerable.Range(0, grainCount)
+                .Select(i => RandomTestGrain(transactionTestGrainClassName))
+                .ToList();
+            var tpsBeforeInterruption = await RecordTPS(txGrains, opsPerGrain);
+            this.logger.LogInformation($"TPS before interruption is {tpsBeforeInterruption}");
+            if(killSiloWhichRunsTm)
+                await KillSilo(location => location.ActiveInCurrentSilo);
+            else
+            {
+                await KillSilo(location => !location.ActiveInCurrentSilo);
+            }
+            await TestingUtils.WaitUntilAsync(lastTry => CheckTPS(txGrains, tpsBeforeInterruption, opsPerGrain, lastTry), RecoveryTimeout);
```

Comment:
Why are we checking TPS? To ensure we recover, we need ensure all transaction operations begin to succeed, so checking for no failures seems more strait forward.  Load tests will vet this behavior under load.

Reply:
I disagree. No failures = recovered seems harsh, since aborts is considered normal behavior?

Reply:
Why should aborts be considered a normal behavior?  In cases were there are conflicts, it may (should not be but for now may) be considered normal, but we can structure the transaction so they do not conflict.